src/engine.rs

```diff
@@ -121,17 +121,25 @@ impl Engine {
     }
 
     /// Increases the balance for a given address.
-    pub fn increase_balance(address: &Address, amount: &U256) {
-        let mut balance = Self::get_balance(address);
-        balance += *amount;
-        Self::set_balance(address, &balance);
+    fn increase_balance(address: &Address, amount: &U256) -> Result<(), ExitError> {
+        let balance = Self::get_balance(address);
+        let (new_balance, overflow) = balance.overflowing_add(*amount);
+        if overflow {
+            return Err(ExitError::StackOverflow);
+        }
+        Self::set_balance(address, &new_balance);
+        Ok(())
     }
 
     /// Decreases the balance for a given address.
-    pub fn decrease_balance(address: &Address, amount: &U256) {
-        let mut balance = Self::get_balance(address);
-        balance -= *amount;
-        Self::set_balance(address, &balance);
+    fn decrease_balance(address: &Address, amount: &U256) -> Result<(), ExitError> {
+        let balance = Self::get_balance(address);
+        let (new_balance, underflow) = balance.overflowing_sub(*amount);
+        if underflow {
+            return Err(ExitError::StackUnderflow);
```

Comment:
This is not a stack underflow, I think we should use `ExitError::Other` (see also above comment).